.github/scripts/install-slices

```diff
@@ -0,0 +1,324 @@
+    try:
+        p.package = data["package"]
+        p.slices = []
+        for key in data["slices"]:
+            p.slices.append(key)
+        p.slices = sorted(p.slices)
+    except KeyError as e:
+        logging.error(f"{filepath}: key {e} not found")
+        exit(1)
+    except Exception as e:
+        logging.error(f"{filepath}: {e}")
+        exit(1)
+    return p
+
+
+def query_package_existence(
+    packages: list[str],
+    arch: list[str] = [],
+    components: list[str] = [],
+    suites: list[str] = [],
+) -> tuple[list[str], list[str]]:
+    """
+    Check which packages exist in the archive. Return a list of packages
+    that exist and another list for which do not.
+    """
+    # prepare cmd
+    args = ["rmadison"]
+    if len(arch) > 0:
+        args += ["--architecture", ",".join(arch)]
+    if len(components) > 0:
+        args += ["--component", ",".join(components)]
+    if len(suites) > 0:
+        args += ["--suite", ",".join(suites)]
+    args.append(" ".join(packages))
+    # query the archives using rmadison
+    logging.debug("Querying the archives for packages...")
+    logging.debug("Executing " + " ".join(args))
+    res = subprocess.run(args, capture_output=True, text=True)
+    if res.returncode != 0:
+        logging.error(f"Failed to query the archives {res.returncode}")
+        exit(res.returncode)
+    output = res.stdout.rstrip()
+    logging.debug(f"Archive query output:\n{output}")
+    # parse the output for available packages
+    available = {}
+    for line in output.split("\n"):
+        line = line.strip()
+        if line == "":
+            continue
+        pkg = line.split("|")[0].strip()
+        available[pkg] = True
+    found = []
+    missing = []
+    for p in packages:
+        if available.get(p):
+            found.append(p)
+        else:
+            missing.append(p)
+    return found, missing
+
+
+def ensure_package_existence(packages: list[str], archive: Archive) -> None:
+    """
+    Ensure that packages exist in the archive for any arch.
+    """
+    _, missing = query_package_existence(
+        packages,
+        components=archive.components,
+        suites=archive.suites,
+    )
+    if len(missing) > 0:
+        logging.error(
+            "The following packages do not exist:\n"
+            + "\n".join(f"  - {p}" for p in missing)
+        )
+        exit(1)
+
+
+def filter_packages(
+    packages: list[Package],
+    arch: str,
+    release: str,
+    ignore_missing: bool = False,
+) -> tuple[list[Package], list[Package]]:
+    """
+    Filter the Packages using the following criteria:
+        - if "ignore_missing" is true, ignore package if it does not exist
+          in the archive for [arch, release].
+    """
+    ignored = []
+    if ignore_missing:
+        package_names = [p.package for p in packages]
+        archive = parse_archive(release)
+        # ensure that packages exist in the archive.
+        ensure_package_existence(package_names, archive)
```

Comment:
this was the problem I was trying to sort out during our call -> this fn is arch-agnostic, but the script is being called for every matrix in the GH job. We're then re-doing the check multiple times, and possibly failing with the same error for every cell of the matrix

Reply:
We decided that although this means re-doing the check multiple times, and possibly failing with the same error for each arch in a particular release, we will go ahead with this for now and optimise/improve it in the future as needed.

Reply:
Added a comment here: https://github.com/canonical/chisel-releases/pull/119/commits/bbb4c82a26d61da9f93b0eaa69577871eb7ba96c